### PR TITLE
Fix the bug where "kubectl describe svc' endpoints"behaves unexpectedly.

### LINKED ISSUE #126922: describe service endpoints shows endpoints that are not ready
### What happened?

After upgrading from 1.29 to 1.30.2 `kubectl describe service` in the endpoints section shows all endpoints even if pods are not ready and `kubectl get endpoints` is not showing the endpoint.
Before upgrading, all pods that are not ready were not shown in the endpoints section.

### What did you expect to happen?

IP of pods that are not ready should not be shown in `kubectl describe service` Endpoints section.

### How can we reproduce it (as minimally and precisely as possible)?

Make pod readyness probe will fail:
```kubectl get pods pkad-health -o wide
NAME          READY   STATUS    RESTARTS      AGE    IP          NODE             NOMINATED NODE   READINESS GATES
pkad-health   0/1     Running   2 (94m ago)   119m   10.1.0.19   docker-desktop   <none>           <none>
```

In the service describe Endpoints section it should not be visable
```
kubectl describe svc pkad
Name:                     pkad
Namespace:                default
Labels:                   <none>
Annotations:              <none>
Selector:                 szkolenie=k8s
Type:                     ClusterIP
IP Family Policy:         SingleStack
IP Families:              IPv4
IP:                       10.106.116.19
IPs:                      10.106.116.19
Port:                     <unset>  8080/TCP
TargetPort:               8080/TCP
Endpoints:                10.1.0.19:8080
Session Affinity:         None
Internal Traffic Policy:  Cluster
Events:                   <none>


kubectl get endpoints pkad
NAME   ENDPOINTS   AGE
pkad               121m

```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.31.0
Kustomize Version: v5.4.2
Server Version: v1.30.2
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

tested on docker desktop and DigitalOcean Kuberentes service.

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by zhifei92**
#### What type of PR is this?
/kind bug

https://github.com/kubernetes/kubernetes/pull/124598, ServiceDescriber been changed since this pr.

Even though we're using endpointslices, we still need to describe ready endpoints.
#### Which issue(s) this PR fixes: 

Fixes https://github.com/kubernetes/kubernetes/issues/126922

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by aroradaman — COMMENTED** [review]
/lgtm

File: staging/src/k8s.io/kubectl/pkg/describe/describe_test.go
Comment (aroradaman) [inline staging/src/k8s.io/kubectl/pkg/describe/describe_test.go:957]: nit: can we have a ready=false condition in any of the endpoint here for assertion

File: staging/src/k8s.io/kubectl/pkg/describe/describe_test.go
Comment (qingwave) [inline staging/src/k8s.io/kubectl/pkg/describe/describe_test.go:816]: The test name should be more meaningful. And a not ready case is better.

**Review by qingwave — CHANGES_REQUESTED** [review]
Just some minor comments.

**Review by ardaguclu — COMMENTED** [review]
We need first review from sig-network about expected behavior.

File: staging/src/k8s.io/kubectl/pkg/describe/describe_test.go
Comment (zhifei92) [reply staging/src/k8s.io/kubectl/pkg/describe/describe_test.go]: > nit: can we have a ready=false condition in any of the endpoint here for assertion

yes, indeed needed, and it also find a bug.

File: staging/src/k8s.io/kubectl/pkg/describe/describe_test.go
Comment (zhifei92) [reply staging/src/k8s.io/kubectl/pkg/describe/describe_test.go]: > Why is this test renamed?

I didn't rename it, just switched the case positions. How is it now?

File: staging/src/k8s.io/kubectl/pkg/describe/describe_test.go
Comment (zhifei92) [reply staging/src/k8s.io/kubectl/pkg/describe/describe_test.go]: The same as above.

File: staging/src/k8s.io/kubectl/pkg/describe/describe_test.go
Comment (zhifei92) [reply staging/src/k8s.io/kubectl/pkg/describe/describe_test.go]: When the `ready condition` field of an `endpointslices` is nil, it should not be displayed by the describe command.

Comment (zhifei92) [reply staging/src/k8s.io/kubectl/pkg/describe/describe_test.go]: Unit tests have been added.

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go
Comment (zhifei92) [reply staging/src/k8s.io/kubectl/pkg/describe/describe.go]: Before https://github.com/kubernetes/kubernetes/pull/124598, `kubectl describe svc`came from the Addresses field in EndpointSubset, and `Addresses` are marked as ready. We maintain the output of `describe svc` to be the same as before  https://github.com/kubernetes/kubernetes/pull/124598.

**Comment by zhifei92** [comment]
> We need first review from sig-network about expected behavior.

/cc @aojea @danwinship If you have time, please help review it, thanks.

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go
Comment (aojea) [inline staging/src/k8s.io/kubectl/pkg/describe/describe.go:5454]:
```suggestion
		// ready indicates that this endpoint is prepared to receive traffic,
		// according to whatever system is managing the endpoint. A nil value
		// indicates an unknown state. In most cases consumers should interpret this
		// unknown state as ready.
		// More info: vendor/k8s.io/api/discovery/v1/types.go
		isReady := endpointSlices[i].Endpoints[j].Conditions.Ready == nil || *endpointSlices[i].Endpoints[j].Conditions.Ready
```

Comment (zhifei92) [reply staging/src/k8s.io/kubectl/pkg/describe/describe.go:5454]: > endpointSlices[i].Endpoints[j].Conditions.Ready == nil || *endpointSlices[i].Endpoints[j].Conditions.Ready

it should be `endpointSlices[i].Endpoints[j].Conditions.Ready == nil && *endpointSlices[i].Endpoints[j].Conditions.Ready`？

Comment (zhifei92) [reply staging/src/k8s.io/kubectl/pkg/describe/describe.go]: >  // A nil value indicates an unknown state. In most cases consumers should interpret this
    // unknown state as ready.

This is a little weird for me.  Thank you very much @aojea

**Comment by zhifei92** [comment]
@ardaguclu If you have time, could you please help review it again?

**Comment by zhifei92** [comment]
@thockin @ardaguclu could you please help review it again? thanks

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/describe/describe.go:5482]: Would shortcutting the logic if it is not ready be better?
```suggestion
						isReady := endpointSlices[i].Endpoints[k].Conditions.Ready == nil || *endpointSlices[i].Endpoints[k].Conditions.Ready
						if !isReady {
						  						// ready indicates that this endpoint is prepared to receive traffic,
						// according to whatever system is managing the endpoint. A nil value
						// indicates an unknown state. In most cases consumers should interpret this
						// unknown state as ready.
						// More info: vendor/k8s.io/api/discovery/v1/types.go
						continue
						}
```

**Comment by ardaguclu** [comment]
Overall looks good to me. Just dropped one comment.

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go
Comment (zhifei92) [reply staging/src/k8s.io/kubectl/pkg/describe/describe.go:5482]: It looks good.

**Comment by zhifei92** [comment]
@ardaguclu @aroradaman Can we merge this PR now

**Comment by ardaguclu** [comment]
Thank you
/triage accepted
/priority important-long-term
/lgtm
/approve

**Comment by ardaguclu** [comment]
/hold
until integration test is stable

**Pull request merged**
